src/main/java/ca/nines/ise/writer/XMLWriter.java

```diff
@@ -16,179 +16,1303 @@
+		switch (node.getName().toUpperCase()) {
+		case "SHY":
+			xmlStack.empty_element(xmlStack.new_element(xml_name));
+		case "L":
+			xmlStack.end_line();
+			xmlStack.new_line(node);
+			break;
+		case "QLN":
+		case "WLN":
+		case "TLN":
+			xmlStack.new_ms_element(xml_name, node.getAttribute("n"));
+			break;
+		case "RULE":
+			xmlStack.end_line();
+			xmlStack.empty_element(set_attributes(node, e,
+					map_put(new_map(), "n", "l"), null, null));
+			break;
+		case "SPACE":
+			xmlStack.ensure_in_line();
+			Element es = set_attributes(node, e,
+					map_put(new_map(), "n", "l"), new String[] { "setting" },
+					null);
+			xmlStack.empty_element(es);
+			/* add extra space */
+			es.appendChild(new Text(" "));
+			break;
+		case "BR":
+			xmlStack.end_line();
+			xmlStack.new_line(new EmptyNode());
+			break;
+		}
+	}
+	
+	private String trim_whitespace(String str){
+		str.replace(" ", "");
+		str.replace("\t", "");
+		str.replace("\r", "");
+		return str;
+	}
+
+	/**
+	 * Parses a text node for newline characters and splits splits the text node
+	 * into separate nodes placed in their own lines. First node is added to the
+	 * current line (if in one) Last line is not ended
+	 * 
+	 * @param t
+	 *            text node
+	 * @param xmlStack
+	 *            XML stack
+	 * @return true if parsed, false otherwise (in page/no newline characters)
+	 */
+	private void parse_text(String text, XMLStack xmlStack) {
+		/*
+		 * if text is all whitespace or in an element which should not be parsed
+		 * element, don't parse further
+		 */
+		if (trim_whitespace(text).isEmpty() || xmlStack.in_dont_parse()){
+			xmlStack.append_to_line(text);
+			return;
+		}
+		String[] lines = null;
+		lines = text.split("\\n+");
+		/* if the text only contains a newline */
+		if (lines.length == 0){
+			xmlStack.end_line();
+			return;
+		}
+		if (lines[0].equals(""))
+			xmlStack.end_line();
+		/* if not in a line, start one; text can not be the child of work */
+		xmlStack.ensure_in_line();
+		/* if there is no newline character, don't parse further */
+		if (!text.contains("\n")){
+			xmlStack.append_to_line(text);
+			return;
+		}
+		for (int i = 0; i < lines.length; i++) {
+			/* end empty lines */
+			if (normalizeBoundary(lines[i]).isEmpty()){
+				continue;
+			}
+			/* add each line as its own line */
+			xmlStack.ensure_in_line();
+			xmlStack.append_to_line(lines[i]);
+			/* don't end the last line*/
+			if (i < (lines.length - 1))
+				xmlStack.end_line();
+		}
+	}
+
+	/**
+	 * wrapper for creating a HashMap
+	 * 
+	 * @return new HashMap<String,String>
+	 */
+	private HashMap<String, String> new_map() {
+		return new HashMap<String, String>();
+	}
+
+	/**
+	 * Wrapper for putting 2 strings into a HashMap<String,String>
+	 * 
+	 * @param hm
+	 *            HashMap
+	 * @param s1
+	 *            first string
+	 * @param s2
+	 *            second string
+	 * @return HashMap (@hm)
+	 */
+	private HashMap<String, String> map_put(HashMap<String, String> hm,
```

Comment:
Haha. It saves a little bit if you want to write it all in one line (like I did). When you straight put the strings in the map, the map is not returned (so you can't do something like set_attributes(n,e,hm.put(s1,s2)). Honestly, it's probably not very necessary.